File: plldb-server/src/main/java/cn/twincest/plldb/util/Parameter.java
package cn.twincest.plldb.util;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.mongodb.lang.Nullable;
import com.mongodb.lang.NonNull;

import java.util.List;

public class Parameter {

	// field

	public String method;

	public JsonObject data;

	// has

	public boolean has(String key) {
		return this.data.has(key);
	}

	public boolean has(List<String> keyList) {
		for (var key : keyList) {
			if (!this.has(key)) {
				return false;
			}
		}
		return true;
	}

	// get // not undefined

	@NonNull
	public JsonElement get(String key) throws ValueUndefinedException {
		var value = this.data.get(key);
		if (value == null) {
			throw new ValueUndefinedException(key);
		}
		return value;
	}

	@Nullable
	public Boolean getBoolean(String key) throws ValueUndefinedException {
		var value = this.get(key);
		return value.isJsonNull() ? null : value.getAsBoolean();
	}

	@Nullable
	public Integer getInteger(String key) throws ValueUndefinedException {
		var value = this.get(key);
		return value.isJsonNull() ? null : value.getAsInt();
	}

	@Nullable
	public String getString(String key) throws ValueUndefinedException {
		var value = this.get(key);
		return value.isJsonNull() ? null : value.getAsString();
	}

	@Nullable
	public JsonArray getArray(String key) throws ValueUndefinedException {
		var value = this.get(key);
		return value.isJsonNull() ? null : value.getAsJsonArray();
	}

	@Nullable
	public JsonObject getObject(String key) throws ValueUndefinedException {
		var value = this.get(key);
		return value.isJsonNull() ? null : value.getAsJsonObject();
	}

	// get required // not undefined or null

	@NonNull
	public JsonElement getRequired(String key) throws ValueUndefinedException, ValueNullException {
		var value = this.data.get(key);
		if (value == null) {
			throw new ValueUndefinedException(key);
		}
		if (value.isJsonNull()) {
			throw new ValueNullException(key);
		}
		return value;
	}

	@NonNull
	public Boolean getBooleanRequired(String key) throws ValueUndefinedException, ValueNullException {
		var value = this.getRequired(key);
		return value.getAsBoolean();
	}

	@NonNull
	public Integer getIntegerRequired(String key) throws ValueUndefinedException, ValueNullException {
		var value = this.getRequired(key);
		return value.getAsInt();
	}

	@NonNull
	public String getStringRequired(String key) throws ValueUndefinedException, ValueNullException {
		var value = this.getRequired(key);
		return value.getAsString();
	}

	@NonNull
	public JsonArray getArrayRequired(String key) throws ValueUndefinedException, ValueNullException {
		var value = this.getRequired(key);
		return value.getAsJsonArray();
	}

	@NonNull
	public JsonObject getObjectRequired(String key) throws ValueUndefinedException, ValueNullException {
		var value = this.getRequired(key);
		return value.getAsJsonObject();
	}

	// exception

	public static class ValueUndefinedException extends Exception {
		public ValueUndefinedException(String key) {
			super(String.format("parameter data `%s` is undefined", key));
		}
	}

	public static class ValueNullException extends Exception {
		public ValueNullException(String key) {
			super(String.format("parameter data `%s` is null", key));
		}
	}

}

File: plldb-server/src-refactor/main/java/cn/twin_kles/plldb/util/Parameter.kt
package cn.twin_kles.plldb.util

import kotlin.Throws
import com.google.gson.JsonElement
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import java.lang.Exception

class Parameter {

    // field
    var method: String? = null
    var data: JsonObject? = null

    // has

    fun has(key: String): Boolean {
        return data!!.has(key)
    }

    fun has(keyList: List<String>): Boolean {
        for (key in keyList) {
            if (!has(key)) {
                return false
            }
        }
        return true
    }

    // get // not undefined
    @Throws(ValueUndefinedException::class)
    fun get(key: String): JsonElement {
        return data!![key] ?: throw ValueUndefinedException(key)
    }

    @Throws(ValueUndefinedException::class)
    fun getBoolean(key: String): Boolean? {
        val value = get(key)
        return if (value.isJsonNull) null else value.asBoolean
    }

    @Throws(ValueUndefinedException::class)
    fun getInteger(key: String): Int? {
        val value = get(key)
        return if (value.isJsonNull) null else value.asInt
    }

    @Throws(ValueUndefinedException::class)
    fun getString(key: String): String? {
        val value = get(key)
        return if (value.isJsonNull) null else value.asString
    }

    @Throws(ValueUndefinedException::class)
    fun getArray(key: String): JsonArray? {
        val value = get(key)
        return if (value.isJsonNull) null else value.asJsonArray
    }

    @Throws(ValueUndefinedException::class)
    fun getObject(key: String): JsonObject? {
        val value = get(key)
        return if (value.isJsonNull) null else value.asJsonObject
    }

    // get required // not undefined or null
    @Throws(ValueUndefinedException::class, ValueNullException::class)
    fun getRequired(key: String): JsonElement {
        val value = data!![key] ?: throw ValueUndefinedException(key)
        if (value.isJsonNull) {
            throw ValueNullException(key)
        }
        return value
    }

    @Throws(ValueUndefinedException::class, ValueNullException::class)
    fun getBooleanRequired(key: String): Boolean {
        val value = getRequired(key)
        return value.asBoolean
    }

    @Throws(ValueUndefinedException::class, ValueNullException::class)
    fun getIntegerRequired(key: String): Int {
        val value = getRequired(key)
        return value.asInt
    }

    @Throws(ValueUndefinedException::class, ValueNullException::class)
    fun getStringRequired(key: String): String {
        val value = getRequired(key)
        return value.asString
    }

    @Throws(ValueUndefinedException::class, ValueNullException::class)
    fun getArrayRequired(key: String): JsonArray {
        val value = getRequired(key)
        return value.asJsonArray
    }

    @Throws(ValueUndefinedException::class, ValueNullException::class)
    fun getObjectRequired(key: String): JsonObject {
        val value = getRequired(key)
        return value.asJsonObject
    }

    // exception
    class ValueUndefinedException(key: String) : Exception("parameter data `$key` is undefined")
    class ValueNullException(key: String) : Exception("parameter data `$key` is null")

}